### PR TITLE
[Merged by Bors] - feat: the pairing of a Lie algebra root space with its negative

### PR DESCRIPTION
**Pull request by ocfnash**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Logic/Basic.lean [hunk 1]
```diff
@@ -776,6 +776,12 @@ theorem exists_apply_eq (a : α) (b : β) : ∃ f : α → β, f a = b := ⟨fun
   ⟨fun ⟨_, ⟨a, ha, b, hb, hab⟩, hc⟩ ↦ ⟨a, ha, b, hb, hab.symm ▸ hc⟩,
     fun ⟨a, ha, b, hb, hab⟩ ↦ ⟨f a b, ⟨a, ha, b, hb, rfl⟩, hab⟩⟩
 
+@[simp] theorem exists_exists_and_exists_and_eq_and' {α β γ : Type*}
```
Comment (jcommelin) [inline Mathlib/Logic/Basic.lean:779]: I'm not sure I follow the name. Maybe just
```suggestion
@[simp] theorem exists_exists_exists_and_eq {α β γ : Type*}
```

File: Mathlib/Algebra/Lie/Weights/Cartan.lean [hunk 2]
```diff
@@ -263,4 +263,36 @@ theorem rootSpace_zero_eq (H : LieSubalgebra R L) [H.IsCartanSubalgebra] [IsNoet
...
+  exact fun p ↦ (rootSpaceProduct R L H α (-α) 0 (add_neg_self α) p).property)
+
+@[simp]
+lemma coe_rootSpaceProductNegSelf_apply (x : rootSpace H α) (y : rootSpace H (-α)) :
+    (rootSpaceProductNegSelf α (x ⊗ₜ y) : L) = ⁅(x : L), (y : L)⁆ :=
+  rfl
+
+lemma mem_range_rootSpaceProductNegSelf {x : H} :
+    x ∈ (rootSpaceProductNegSelf α).range ↔
+    (x : L) ∈ Submodule.span R {w | ∃ᵉ (y ∈ rootSpace H α) (z ∈ rootSpace H (-α)), ⁅y, z⁆ = w} := by
```
Comment (jcommelin) [inline Mathlib/Algebra/Lie/Weights/Cartan.lean:287]: Nowadays you can even do this!
```suggestion
    (x : L) ∈ Submodule.span R {⁅y, z⁆ | (y ∈ rootSpace H α) (z ∈ rootSpace H (-α))} := by
```

Comment (ocfnash) [reply Mathlib/Algebra/Lie/Weights/Cartan.lean]: Ooooh neat!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**